In [63]:
import numpy as np 
import pandas as pd

In [76]:
df = pd.concat(map(pd.read_csv, ['Brazil.csv', 'germany.csv','germany1.csv', 
                                 'Portugal1.csv', 'Spain1.csv', 'Switzerland.csv',
                                'Switzerland1.csv', 'France1.csv',
                                'Brazil1.csv', 'Netherlands.csv', 
                                'Belgium.csv', 'USA.csv', 'England.csv',
                                'Argentina.csv', 'Uruguay.csv', 
                                'Senegal.csv', 'Ecuador.csv', 'CostaRica.csv',
                                'Mexico.csv', 'France.csv', 'Denmark.csv',
                                'Australia.csv', 'Serbia.csv',
                                'Cameroon.csv', 'Ghana.csv', 'Iran.csv',
                                'Poland.csv', 'SaudI.csv', 'Tunisia.csv', 
                                 'Japan.csv', 'Croatia.csv', 'Morocco.csv',
                                'USA1.csv', 'Spain2.csv', 'Portugal2.csv', 
                                 'England2.csv', 'Belgium2.csv', 'Uruguay2.csv',
                                'Canada2.csv', 'Netherlands2.csv', 'Saudi2.csv', 
                                'Croatia2.csv', 'Cameroon2.csv', 'Denmark2.csv',
                                'Costa2.csv', 'Morocco2.csv', 'Poland2.csv',
                                'Tunisia2.csv', 'Japan2.csv', 'Argentina2.csv']))



In [77]:
df.shape

(158, 17)

In [78]:
df

goals_z      xg_z  crosses_z  boxtouches_z  passes_z  progpasses_z  \
0   0.423077  0.146923  -0.136154     -0.030000  0.429231      0.037692   
1   0.216923  0.348462   0.031538      0.158462  0.835385      0.626923   
2   0.113846  0.392308   0.599231      0.513846  0.833077      0.405385   
3   0.479231  0.609231   0.227692      0.450769  0.770769      0.042308   
4   0.877692  0.773846   0.428462      0.659231  0.754615      0.335385   
..       ...       ...        ...           ...       ...           ...   
0   0.234615  0.405385   0.200769      0.554615  0.370769      0.331538   
1   0.536154  0.170769   0.166154      0.190769 -0.167692     -0.196154   
2  -0.070769 -0.368462  -0.493846     -0.247692 -0.681538     -0.536923   
3   0.420000  0.659231   0.472308      0.616923  0.452308      0.056923   
4   0.472308  0.458462  -0.023846      0.078462  0.416923      0.273846   

    takeons_z  progruns_z  tackles_z  interceptions_z  clearances_z  blocks_z  \
0    0.244615   -0.220000   0.216154         0.270000     -0.076923 -0.097692   
1    0.270000    0.266923   1.143846         0.834615     -0.059231 -0.134615   
2    0.521538    1.139231   1.109231         0.859231      0.084615  0.103077   
3    0.337692    0.927692   0.506923         1.015385      0.020769  0.381538   
4    0.023077    0.638462   0.493846         0.637692     -0.117692 -0.033846   
..        ...         ...        ...              ...           ...       ...   
0    0.350769    0.026923   1.114615         0.497692     -0.392308 -0.084615   
1    0.557692   -0.204615  -0.063846         0.343077      0.177692  0.333846   
2   -0.210000   -0.427692  -0.539231        -0.622308     -0.313846 -0.307692   
3    1.100769    0.366923   0.593846         0.558462     -0.463846 -0.183077   
4    0.693077    0.513077   0.130000         0.178462     -0.239231  0.021538   

    aerials_z   fouls_z  fouled_z    nsxg_z  results  
0   -0.020000 -0.224615  0.100769 -0.124615        2  
1    0.196923 -0.016923  0.033846  0.146923        1  
2    0.683846  0.213846  0.353846  0.786923        9  
3    0.038462  0.039231  0.022308  0.692308        5  
4    0.572308 -0.016154 -0.096923  0.890769        5  
..        ...       ...       ...       ...      ...  
0    0.480000  0.017692  0.514615  0.296154       11  
1    0.128462 -0.034615  0.470000  0.228462        1  
2   -0.236923 -0.200769 -0.405385 -0.526154        2  
3   -0.121538  0.094615  0.826923  0.990769       10  
4   -0.318462  0.204615  0.463077  0.521538        6  

[158 rows x 17 columns]

In [79]:
df.corr()

goals_z      xg_z  crosses_z  boxtouches_z  passes_z  \
goals_z          1.000000  0.593882   0.042906      0.300056  0.257896   
xg_z             0.593882  1.000000   0.383316      0.669005  0.379186   
crosses_z        0.042906  0.383316   1.000000      0.644555  0.234163   
boxtouches_z     0.300056  0.669005   0.644555      1.000000  0.382181   
passes_z         0.257896  0.379186   0.234163      0.382181  1.000000   
progpasses_z     0.104536  0.168393   0.285101      0.238482  0.743879   
takeons_z        0.292063  0.341023   0.115226      0.202332  0.398349   
progruns_z       0.220753  0.439055   0.262945      0.432240  0.656938   
tackles_z        0.313448  0.408050   0.193384      0.396708  0.757077   
interceptions_z  0.262674  0.334599   0.230263      0.337851  0.724294   
clearances_z    -0.169119 -0.356212  -0.191263     -0.420998 -0.118588   
blocks_z         0.080306  0.137321   0.075910      0.233539  0.430210   
aerials_z        0.065206  0.085119   0.264544      0.280895 -0.000897   
fouls_z         -0.154157 -0.080857   0.025607     -0.103904 -0.019146   
fouled_z         0.015971  0.150826   0.202819      0.143837  0.276909   
nsxg_z           0.370490  0.679585   0.666068      0.855952  0.520417   
results         -0.642001 -0.315752   0.034878     -0.200537 -0.178234   

                 progpasses_z  takeons_z  progruns_z  tackles_z  \
goals_z              0.104536   0.292063    0.220753   0.313448   
xg_z                 0.168393   0.341023    0.439055   0.408050   
crosses_z            0.285101   0.115226    0.262945   0.193384   
boxtouches_z         0.238482   0.202332    0.432240   0.396708   
passes_z             0.743879   0.398349    0.656938   0.757077   
progpasses_z         1.000000   0.180785    0.419958   0.523392   
takeons_z            0.180785   1.000000    0.567979   0.388583   
progruns_z           0.419958   0.567979    1.000000   0.592725   
tackles_z            0.523392   0.388583    0.592725   1.000000   
interceptions_z      0.523185   0.377466    0.488475   0.712897   
clearances_z        -0.010195  -0.075201   -0.062295  -0.218618   
blocks_z             0.384167   0.205780    0.358292   0.325432   
aerials_z            0.033427  -0.025027    0.028458   0.048009   
fouls_z              0.164577   0.015762    0.001364  -0.082459   
fouled_z             0.117262   0.369726    0.301723   0.286030   
nsxg_z               0.395564   0.373965    0.556932   0.498909   
results             -0.084719  -0.174466   -0.201809  -0.237599   

                 interceptions_z  clearances_z  blocks_z  aerials_z   fouls_z  \
goals_z                 0.262674     -0.169119  0.080306   0.065206 -0.154157   
xg_z                    0.334599     -0.356212  0.137321   0.085119 -0.080857   
crosses_z               0.230263     -0.191263  0.075910   0.264544  0.025607   
boxtouches_z            0.337851     -0.420998  0.233539   0.280895 -0.103904   
passes_z                0.724294     -0.118588  0.430210  -0.000897 -0.019146   
progpasses_z            0.523185     -0.010195  0.384167   0.033427  0.164577   
takeons_z               0.377466     -0.075201  0.205780  -0.025027  0.015762   
progruns_z              0.488475     -0.062295  0.358292   0.028458  0.001364   
tackles_z               0.712897     -0.218618  0.325432   0.048009 -0.082459   
interceptions_z         1.000000     -0.099396  0.341119   0.064151  0.023685   
clearances_z           -0.099396      1.000000  0.184999   0.211456  0.278011   
blocks_z                0.341119      0.184999  1.000000   0.099023  0.213226   
aerials_z               0.064151      0.211456  0.099023   1.000000  0.153951   
fouls_z                 0.023685      0.278011  0.213226   0.153951  1.000000   
fouled_z                0.314982      0.122941  0.143722   0.011390  0.228028   
nsxg_z                  0.458088     -0.420326  0.248688   0.146909 -0.044694   
results                -0.166157      0.089839 -0.103182  -0.071186  0.166183   

                 fo

In [80]:
df.shape

(158, 17)

In [92]:
# df.to_csv('WC-stats.csv', index=False)


In [82]:
from sklearn.linear_model import LinearRegression
from sklearn.inspection import permutation_importance
from sklearn.model_selection import train_test_split
from sklearn.linear_model import Ridge



In [83]:
df.head()

goals_z      xg_z  crosses_z  boxtouches_z  passes_z  progpasses_z  \
0  0.423077  0.146923  -0.136154     -0.030000  0.429231      0.037692   
1  0.216923  0.348462   0.031538      0.158462  0.835385      0.626923   
2  0.113846  0.392308   0.599231      0.513846  0.833077      0.405385   
3  0.479231  0.609231   0.227692      0.450769  0.770769      0.042308   
4  0.877692  0.773846   0.428462      0.659231  0.754615      0.335385   

   takeons_z  progruns_z  tackles_z  interceptions_z  clearances_z  blocks_z  \
0   0.244615   -0.220000   0.216154         0.270000     -0.076923 -0.097692   
1   0.270000    0.266923   1.143846         0.834615     -0.059231 -0.134615   
2   0.521538    1.139231   1.109231         0.859231      0.084615  0.103077   
3   0.337692    0.927692   0.506923         1.015385      0.020769  0.381538   
4   0.023077    0.638462   0.493846         0.637692     -0.117692 -0.033846   

   aerials_z   fouls_z  fouled_z    nsxg_z  results  
0  -0.020000 -0.224615  0.100769 -0.124615        2  
1   0.196923 -0.016923  0.033846  0.146923        1  
2   0.683846  0.213846  0.353846  0.786923        9  
3   0.038462  0.039231  0.022308  0.692308        5  
4   0.572308 -0.016154 -0.096923  0.890769        5

In [84]:
y = df['results']

In [85]:
X = df.drop(columns=['results'])

In [86]:
X_train, X_test, y_train, y_test = train_test_split(
     X, y, test_size=0.20, random_state=42)

In [87]:
model = Ridge(alpha=1e-2).fit(X_train, y_train)
model.score(X, y)

0.42629810615686115

In [88]:
r = permutation_importance(model, X, y, n_repeats=30, random_state=0)

# Permutation feature importance

In [90]:
scoring = ['r2', 'neg_mean_absolute_percentage_error', 'neg_mean_squared_error']
r_multi = permutation_importance(
     model, X, y, n_repeats=30, random_state=0, scoring=scoring)

for metric in r_multi:
     print(f"{metric}") 
     r = r_multi[metric]
     for i in r.importances_mean.argsort()[::-1]:
            if r.importances_mean[i] - 2 * r.importances_std[i] > 0:
                print(f"    {X.columns[i]:<8}"
                      f"{r.importances_mean[i]:.3f}"
                      f" +/- {r.importances_std[i]:.3f}")
                
                

r2
    goals_z 1.282 +/- 0.149
    progruns_z0.096 +/- 0.029
    xg_z    0.095 +/- 0.030
    takeons_z0.026 +/- 0.013
neg_mean_absolute_percentage_error
    goals_z 0.988 +/- 0.115
neg_mean_squared_error
    goals_z 108.981 +/- 12.672
    progruns_z8.139 +/- 2.494
    xg_z    8.047 +/- 2.509
    takeons_z2.195 +/- 1.067


- df.shape = (158, 17)
- Ridge Regression = 0.42629810615686115
- r2
    - goals_z 1.282 +/- 0.149
    - progruns_z0.096 +/- 0.029
    - xg_z    0.095 +/- 0.030
    - takeons_z0.026 +/- 0.013
- neg_mean_absolute_percentage_error
    - goals_z 0.988 +/- 0.115
- neg_mean_squared_error
    - goals_z 108.981 +/- 12.672
    - progruns_z8.139 +/- 2.494
    - xg_z    8.047 +/- 2.509
    - takeons_z2.195 +/- 1.067


- df.shape = (125, 17)

- Ridge Regression = 0.4494483686544971

- r2
    - goals_z 0.982 +/- 0.123
    - progruns_z0.132 +/- 0.039
    - xg_z    0.056 +/- 0.017
    - boxtouches_z0.046 +/- 0.023
    - nsxg_z  0.039 +/- 0.013
    - aerials_z0.031 +/- 0.014
- neg_mean_absolute_percentage_error
    - goals_z 0.888 +/- 0.122
    - progruns_z0.181 +/- 0.059
- neg_mean_squared_error
    - goals_z 89.125 +/- 11.206
    - progruns_z11.964 +/- 3.505
    - xg_z    5.055 +/- 1.543
    - boxtouches_z4.180 +/- 2.062
    - nsxg_z  3.509 +/- 1.182
    - aerials_z2.782 +/- 1.244

- df.shape = (145, 17)
- Ridge Regression = 0.43985733944194594

- r2
    - goals_z 0.919 +/- 0.122
    - progruns_z0.077 +/- 0.028
    - nsxg_z  0.069 +/- 0.026
    - boxtouches_z0.034 +/- 0.015
- neg_mean_absolute_percentage_error
    - goals_z 0.817 +/- 0.115
    - progruns_z0.124 +/- 0.036
    - tackles_z0.060 +/- 0.022
    - boxtouches_z0.048 +/- 0.023
- neg_mean_squared_error
    - goals_z 78.467 +/- 10.375
    - progruns_z6.549 +/- 2.364
    - nsxg_z  5.854 +/- 2.186
    - boxtouches_z2.943 +/- 1.292

- NMAPE = the higher the better
- NMSE
- R2 =  is a statistical measure of how close the data are to the fitted regression line . It is the percentage of variation (from 0 to 1) explained by the relationship between two variables.. metric that is used to evaluate the performance of a regression-based machine learning model.